### PR TITLE
GH-3: Make HTTP Source SI-based, add ConfigProps

### PR DESCRIPTION
**Pull request by artembilan**
Fixes spring-cloud-stream-app-starters/http#3
      spring-cloud-stream-app-starters/http#1

Since we can't use `@ConfigurationProperties` with the `@RequestMapping`, plus right now there is no way to pass inbound headers,
rework the source to use Spring Integration HTTP Inbound Channel Adapter

### CODE REVIEW

**Comment by artembilan** [comment]
Note: I wasn't able to modify `README` for new configuration properties because the `spring-cloud-app-starter-doc-maven-plugin`  doesn't work for me properly on Windows.

I have this debug on the matter:
```
[DEBUG] Resolved classpath location [/META-INF/spring-configuration-metadata-whitelist.properties] to resources [URL [jar:file:/D:/SpringIO/spring-cloud-stream-app-starters/http/spring-cloud-starter-stream-source-http/target/spring-cloud-starter-stream-source-http-1.2.0.BUILD-SNAPSHOT.jar!/META-INF/spring-configuration-metadata-whitelist.properties], URL [file:/D:/SpringIO/spring-cloud-stream-app-starters/http/spring-cloud-starter-stream-source-http/target/classes/META-INF/spring-configuration-metadata-whitelist.properties]]
[DEBUG] Resolved classpath location [/META-INF/spring-configuration-metadata.json] to resources [URL [jar:file:/D:/SpringIO/spring-cloud-stream-app-starters/http/spring-cloud-starter-stream-source-http/target/spring-cloud-starter-stream-source-http-1.2.0.BUILD-SNAPSHOT.jar!/META-INF/spring-configuration-metadata.json], URL [file:/D:/SpringIO/spring-cloud-stream-app-starters/http/spring-cloud-starter-stream-source-http/target/classes/META-INF/spring-configuration-metadata.json], URL [jar:file:/C:/Users/abilan/.m2/repository/org/springframework/cloud/spring-cloud-stream/1.2.0.BUILD-SNAPSHOT/spring-cloud-stream-1.2.0.BUILD-SNAPSHOT.jar!/META-INF/spring-configuration-metadata.json], URL [jar:file:/C:/Users/abilan/.m2/repository/org/springframework/boot/spring-boot-actuator/1.5.0.RC1/spring-boot-actuator-1.5.0.RC1.jar!/META-INF/spring-configuration-metadata.json], URL [jar:file:/C:/Users/abilan/.m2/repository/org/springframework/boot/spring-boot-autoconfigure/1.5.0.RC1/spring-boot-autoconfigure-1.5.0.RC1.jar!/META-INF/spring-configuration-metadata.json], URL [jar:file:/C:/Users/abilan/.m2/repository/org/springframework/boot/spring-boot/1.5.0.RC1/spring-boot-1.5.0.RC1.jar!/META-INF/spring-configuration-metadata.json]]
[DEBUG] Documenting http.mapped-request-headers
[DEBUG] Documenting http.path-pattern
[DEBUG] Documenting server.port
[INFO] Documented 3 configuration properties
```
But nothing has written to the `README` back.

Or I am missing something else, or `spring-cloud-app-starter-doc-maven-plugin` doesn't work on Windows.

Thanks

/CC @ericbottard

File: spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java [hunk 1]
```diff
@@ -41,28 +39,50 @@
...
-				new MessageHeaders(Collections.singletonMap(MessageHeaders.CONTENT_TYPE, contentType))));
+	private HttpRequestHandlerEndpointSpec buildHttpRequestHandlerEndpointSpec(final String... consumes) {
+		// TODO Until SI Java DSL 1.2.2. Use mappedRequestHeaders() instead
+		DefaultHttpHeaderMapper defaultHttpHeaderMapper = DefaultHttpHeaderMapper.inboundMapper();
+		defaultHttpHeaderMapper.setInboundHeaderNames(this.properties.getMappedRequestHeaders());
+
+		return Http.inboundChannelAdapter(this.properties.getPathPattern())
+				.headerMapper(defaultHttpHeaderMapper)
+//				.mappedRequestHeaders(this.properties.getMappedRequestHeaders())
+				.statusCodeExpression(new ValueExpression<>(HttpStatus.ACCEPTED))
```
Comment (garyrussell) [inline spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java:75]: Shouldn't we also provide `.headerExpressions(...)` ??

e.g. to add headers from `#pathVariables`

**Review by garyrussell — CHANGES_REQUESTED** [review]

File: spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java (hunk 1 above)
Comment (garyrussell) [reply spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java:75]: Also `payloadExpression` and `statusCodeExpression`.

Comment (artembilan) [reply spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java:75]: That is a story for future.
Right now I fix only bugs and try to keep an existing functionality

File: http-app-dependencies/pom.xml [hunk 2]
```diff
@@ -1,93 +1,29 @@
...
-                <groupId>org.springframework.cloud.stream.app</groupId>
-                <artifactId>spring-cloud-starter-stream-source-http</artifactId>
-                <version>1.2.0.BUILD-SNAPSHOT</version>
-            </dependency>
-        </dependencies>
-    </dependencyManagement>
-<profiles>
-		<profile>
-			<id>spring</id>
-			<repositories>
```
Comment (markpollack) [inline http-app-dependencies/pom.xml:30]: Why did you remove all these repositories?  You can't pick up the repositories defined in the parent if you can't find the parent.

Comment (markpollack) [reply http-app-dependencies/pom.xml:30]: never mind....i  thought this was the root pom

Comment (artembilan) [reply http-app-dependencies/pom.xml:30]: Now this repos don't work at all.
We use this `dependencies` artifact in the target app as:
```
<dependency>
        <groupId>org.springframework.cloud.stream.app</groupId>
        <artifactId>http-app-dependencies</artifactId>
        <version>1.2.0.BUILD-SNAPSHOT</version>
        <type>pom</type>
        <scope>import</scope>
</dependency>
```
I'm not sure what you mean about the parent issue, but here `http-app-starters-build` is parent on place in this very project with proper repos definitions. So, what is the concern ?

File: spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java (hunk 1 above)
Comment (garyrussell) [reply spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java:75]: But you claim in the comment that it resolves spring-cloud-stream-app-starters/http#1
and it doesn't - that story is for a fully-featured SI-based source.

Comment (artembilan) [reply spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java:75]: OK. I can wash away that part 😄 
To be honest I'm not sure yet what else we can do here...
I can add those `headerExpressions` and `payloadExpression` as well as `statusCodeExpression`, but not sure how that correlates with the current `String/byte[]` distinguishing functionality...

Comment (mbogoevici) [reply spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java:75]: I am wondering why do we needed that routing functionality in the first place, as publishing to the binder would reconvert the String to a byte[] anyway. Also, because the latter conversion is UTF-8 based and we send the header unchanged, doesn't it mean that we are republishing UTF-8 bytes but potentially advertise a different encoding (e.g. if the original one was `UTF-32` or something else)?

Comment (artembilan) [reply spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java:75]: Sorry, I am somehow useless for this discussion: don't feel confident to argue.
I can agree and just make a single adapters with all the request options from Gary.
We can add `requestPayloadType` as well, if there is such a requirements for target use-case.
The `payloadExpression` then can be used to extract from that object desired properties.
And only that result will be send to downstream.

I wonder if we should support other non-GET HTTP methods as well, e.g. PUT.

**Comment by markpollack** [comment]
Oops, I ended up merging this as I didn't see the comments.  It doesn't seem that there was a conclusion to the above discussion...

Regenerated the app to update the README.adoc with new config props.

I tested posting to  `/data` and was also able to access the `/management/env` endpoint.

Let's open a new issue for feature enhancements.

squashed and merged as d38cfb1ca7141529ed24324a9459c5682739e6bf
